File: swift-package/SkylogKit/Sources/SkylogKit/Skylog.swift
/// Global logging facade.
///
/// Mirrors `dev.viethung.skylog.Skylog` from `:skylog-core`.
///
/// `Skylog` is a case-less enum used as a namespace — it cannot be instantiated.
/// All calls delegate to a shared `Logger` instance configured via `Skylog.configure(_:)`.
///
/// **Initialization:** the default `Logger` starts with an empty writer list. Call
/// `Skylog.configure { $0.writers.append(OsLogWriter()) }` at app startup to add writers.
///
/// **Test setup recipe:** tests that need isolated writers should call
/// `Skylog.configure { $0.writers = [fakeWriter] }` before the first log call,
/// or construct a fresh `Logger(config:)` directly instead of using `Skylog`.
public enum Skylog {

    // MARK: - Private shared state

    static let shared = Logger()

    // MARK: - Severity entry points

    /// Log at VERBOSE level.
    public static func v(tag: String? = nil, throwable: Error? = nil, _ message: @autoclosure () -> String) {
        shared.v(tag: tag, throwable: throwable, message())
    }

    /// Log at DEBUG level.
    public static func d(tag: String? = nil, throwable: Error? = nil, _ message: @autoclosure () -> String) {
        shared.d(tag: tag, throwable: throwable, message())
    }

    /// Log at INFO level.
    public static func i(tag: String? = nil, throwable: Error? = nil, _ message: @autoclosure () -> String) {
        shared.i(tag: tag, throwable: throwable, message())
    }

    /// Log at WARN level.
    public static func w(tag: String? = nil, throwable: Error? = nil, _ message: @autoclosure () -> String) {
        shared.w(tag: tag, throwable: throwable, message())
    }

    /// Log at ERROR level.
    public static func e(tag: String? = nil, throwable: Error? = nil, _ message: @autoclosure () -> String) {
        shared.e(tag: tag, throwable: throwable, message())
    }

    /// Log at ASSERT level.
    public static func a(tag: String? = nil, throwable: Error? = nil, _ message: @autoclosure () -> String) {
        shared.a(tag: tag, throwable: throwable, message())
    }

    // MARK: - Structured fields overloads

    /// Log at INFO level with structured key–value fields.
    public static func i(tag: String? = nil, fields: [String: String], throwable: Error? = nil, _ message: @autoclosure () -> String) {
        shared.i(tag: tag, fields: fields, throwable: throwable, message())
    }

    /// Log at DEBUG level with structured key–value fields.
    public static func d(tag: String? = nil, fields: [String: String], throwable: Error? = nil, _ message: @autoclosure () -> String) {
        shared.d(tag: tag, fields: fields, throwable: throwable, message())
    }

    // MARK: - Configuration

    /// Atomically update the shared logger's configuration.
    ///
    /// Example:
    /// ```swift
    /// Skylog.configure {
    ///     $0.minSeverity = .debug
    ///     $0.writers.append(OsLogWriter())
    ///     $0.writers.append(inMemoryWriter)
    /// }
    /// ```
    public static func configure(_ block: (inout SkylogConfig) -> Void) {
        shared.configure(block)
    }

    /// Convenience — set the minimum severity on the shared logger.
    public static func minSeverity(_ severity: Severity) {
        configure { $0.minSeverity = severity }
    }
}

File: skylog-core/src/commonMain/kotlin/dev/viethung/skylog/Skylog.kt
package dev.viethung.skylog

import dev.viethung.skylog.platform.platformDefaultWriter

/**
 * Global logging facade backed by a singleton [Logger].
 *
 * For non-global instances (different tag / writer set), construct a [Logger] directly.
 *
 * ## Initialization contract
 *
 * The `default` [Logger] is initialized eagerly on the first access to `Skylog` (Kotlin
 * `object` semantics). This triggers [platformDefaultWriter] exactly once per process.
 * Safe under Kotlin 2.x's new memory model — [Logger.config] is `@Volatile` and mutated
 * under `synchronized(lock)`.
 *
 * **Test setup recipe:** tests that need to swap in a fake writer must call
 * `Skylog.configure { writers.clear(); writers += fakeWriter }` BEFORE the first
 * `Skylog.v/d/i/w/e/a/configure` call, OR construct a fresh
 * `Logger(SkylogConfig().apply { writers += fakeWriter })` directly instead of using `Skylog`.
 */
object Skylog {
    private val default: Logger = Logger(SkylogConfig().apply { writers += platformDefaultWriter() })

    fun v(tag: String? = null, throwable: Throwable? = null, message: () -> String) =
        default.v(tag, throwable, message)

    fun d(tag: String? = null, throwable: Throwable? = null, message: () -> String) =
        default.d(tag, throwable, message)

    fun i(tag: String? = null, throwable: Throwable? = null, message: () -> String) =
        default.i(tag, throwable, message)

    fun w(tag: String? = null, throwable: Throwable? = null, message: () -> String) =
        default.w(tag, throwable, message)

    fun e(tag: String? = null, throwable: Throwable? = null, message: () -> String) =
        default.e(tag, throwable, message)

    fun a(tag: String? = null, throwable: Throwable? = null, message: () -> String) =
        default.a(tag, throwable, message)

    // Structured fields surface
    fun v(tag: String? = null, fields: Map<String, String>, throwable: Throwable? = null, message: () -> String) =
        default.v(tag, fields, throwable, message)

    fun d(tag: String? = null, fields: Map<String, String>, throwable: Throwable? = null, message: () -> String) =
        default.d(tag, fields, throwable, message)

    fun i(tag: String? = null, fields: Map<String, String>, throwable: Throwable? = null, message: () -> String) =
        default.i(tag, fields, throwable, message)

    fun w(tag: String? = null, fields: Map<String, String>, throwable: Throwable? = null, message: () -> String) =
        default.w(tag, fields, throwable, message)

    fun e(tag: String? = null, fields: Map<String, String>, throwable: Throwable? = null, message: () -> String) =
        default.e(tag, fields, throwable, message)

    fun a(tag: String? = null, fields: Map<String, String>, throwable: Throwable? = null, message: () -> String) =
        default.a(tag, fields, throwable, message)

    fun configure(block: SkylogConfig.() -> Unit) = default.configure(block)

    fun minSeverity(severity: Severity) = configure { minSeverity = severity }
}
